### PR TITLE
[SOFT-396] General power distribution update to rev 3.1

### PR DESCRIPTION
**Pull request by copperium**
This is somewhere between a pin update, a refactor, and a rewrite of power distribution, and definitely shouldn't all be lumped into one PR. But on the upside, PD should be much more pleasant to work with now.

~~Note: just realized one mistake - horn and fan on UV cutoff are BTS7040s so we need to use the BTS7040 driver to get fault handling.~~
*Update: realized that we don't have direct current sense of horn and fan so we have no way of telling if we're in a fault since faults are signalled by very high current sense, so there's no point in using the BTS7040 driver. Guess we'll just have to hope that faults don't happen there? Otherwise the fan/horn will be locked off until we turn them off and on again. But in any case there's nothing we can do about that in firmware.*

## Architectural changes - `output` module

TL;DR we take all the hardware nonsense and wrap it up in `output` and `output_config` modules so the rest of PD doesn't have to deal with it.

The largest architectural change is the introduction of the `output` module. It defines an `Output` type, which is an abstraction of anything that PD can turn on and off and/or measure current from. Outputs can be implemented through BTS7200s, BTS7040s, or GPIO, and the module handles dealing with the multiple implementations and owns all the BTS7200/7040 storages. The goal of the module is to encapsulate all the complexity required to deal with the underlying hardware and wrap it up into two functions which work on (almost) any output:
```c
// Set whether the output is on or off.
StatusCode output_set_state(Output output, OutputState state);

// Read the current that the output is drawing into |*current| in mA.
StatusCode output_read_current(Output output, uint16_t *current);
```
All other PD modules have been rewritten to work in terms of outputs exclusively, which simplifies things a lot!

All the hardware configuration is done in `output_config.c`, which describes how outputs are implemented on hardware.

Some design decisions and notes:
* Outputs don't overlap between front and rear PD, and are specified using a unified `OutputConfig`. This simplifies things and means that output numbers in logs are unambiguous if we don't know which board they came from.
* `OutputConfig` consists of a map of outputs to `OutputSpec`s, which is a tagged union between `OutputBts7200Spec`, `OutputBts7040Spec`, and `OutputGpioSpec`. Each of those types contain the fields needed to implement setting state and reading current.
* We specify information about BTS7200s outside the main `OutputConfig` in `OutputBts7200Info` structs, and then use pointers to them in `OutputBts7200Spec` to associate outputs with BTS7200 info, which reduces duplication. We detect whether two outputs are on the same BTS7200 by whether they use the same pointer - this is the best way I could think of to implement this but I'm open to a better solution.
  * Even though front and rear share the same BTS7200 pins, we use separate `OutputBts7200Info` structs because we still use the #defines in pin_defs.h to give front/rear-specific names to the pins.
* GPIO outputs can't read current because we don't currently have any that need to.
* Because outputs fulfill the same function as `PowerDistributionCurrent` of uniquely identifying a current measurement source, we remove currents.h and just use outputs for the same purpose.
* We no longer read currents from BTS7200s by reading one channel, then the other channel; instead we read one channel at a time (I think this is fine and maybe beneficial). So we update the BTS7200 driver (and the BTS7200 smoke test) so that bts7200_get_measurement only reads from one channel.

## Changes to update to rev 3.1
* On front PD, we use the GPIO FAN output on UV cutoff rather than the FAN_1 and FAN_2 outputs on the BTS7200 on PD.
* Horn was changed to a GPIO output. It's also on UV cutoff.
* We add a UV_VBAT output to read current from the battery on UV cutoff - it doesn't have a corresponding pin to set on/off so we use a hack with a `NONEXISTENT_PCA9539R_PIN`.
* Lots of pins were moved around.
* Confirmed rear fan control PWM ports with hardware, DCDC is on PWM port 1 and enclosure ventilation is on ADT7476A PWM port 3 (= PD PWM output 2, see the schematic lol).

## Functional changes
* Removed all the extraneous `SYSTEM_CAN_MESSAGE_FRONT_POWER` stuff in can_rx_event_mapper_config and pd_gpio_config (and remove the now-unused CAN message itself). It was for "manual control of front pd outputs" but it wasn't used and I don't think it's valuable, and the CAN message definition didn't make sense. Also removed the exported enums referencing it since they're obsolete.
  * If it would be helpful for debugging we could add a message that enables/disables an arbitrary output.
* We check all the status codes in main.c and print a helpful message if any of them return non-OK. Also, we can force either front or rear power distribution in main.c.
* Removed the `ack` field in can_rx_event_mapper since it turns out that *doesn't actually do anything*.
* Added fields to the FRONT_PD_FAULT and REAR_PD_FAULT messages for txing when there's a BTS7200 or BTS7040 fault. To do this, I had to remove the reference voltage field from FRONT_PD_FAULT, but I don't think that was valuable anyways.
* Lengthened the publish_data `ITERATION_WAIT_MS` to try to avoid overwhelming the event queue.
* can_rx_event_mapper/pd_gpio events now use `EVENT_PRIORITY_HIGH` (and the BPS strobe events use `EVENT_PRIORITY_HIGHEST`) to avoid being on the same event queue as CAN TX/RX events from publish_data.
* A bunch of divide-by-zero safeguards in pd_fan_ctrl.c which are probably physically impossible, but who cares about physics.

## Stylistic changes
* Get rid of all instances of the terrible no-good very bad `POWER_DISTRIBUTION` prefix, now all our identifiers are shorter. It's either changed to `PD` or removed altogether.

## Ideas I had but didn't do (yet)
* Merge the current_measurement and publish_data configs somehow since their lists of outputs are identical
* A macro for producing a compound literal array and its length so that we don't have to manually count the elements in every array in a config

We also saved ~1500 bytes of stack space over the old version (6492B vs 7924B), largely due to not allocating as many BTS7200/BTS7040 storages.

### CODE REVIEW

**Review by jessm — APPROVED** [review]
LGTM! Nothing blocking, but I'd appreciate your thoughts / answers to questions 🙂

File: libraries/ms-drivers/src/bts7200_load_switch.c [hunk 1]
```diff
@@ -188,49 +168,30 @@ static void prv_convert_voltage_to_current(Bts7200Storage *storage, uint16_t *me
...
-      (storage->min_fault_voltage_mv <= *meas0) && (*meas0 <= storage->max_fault_voltage_mv);
-  bool fault1 =
-      (storage->min_fault_voltage_mv <= *meas1) && (*meas1 <= storage->max_fault_voltage_mv);
-
-  prv_convert_voltage_to_current(storage, meas0);
-  prv_convert_voltage_to_current(storage, meas1);
-
-  if (fault0 || fault1) {
-    // Only call fault cb if it's not NULL
+  bool fault = storage->min_fault_voltage_mv <= *meas && *meas <= storage->max_fault_voltage_mv;
```
[the remark this answers is not in the document]
Comment (copperium) [reply libraries/ms-drivers/src/bts7200_load_switch.c:185]: I think the original idea here might have been to handle possible integer overflow? In any case it doesn't make much sense, I'll remove the max voltage

File: projects/power_distribution/inc/output_config.h [hunk 2]
```diff
@@ -0,0 +1,5 @@
+#pragma once
+
+#include "output.h"
+
+extern const OutputConfig COMBINED_OUTPUT_CONFIG;
```
[the remark this answers is not in the document]
Comment (copperium) [reply projects/power_distribution/inc/output_config.h:5]: I was following the precedent from old PD (which I guess I set), but yeah I think a `g_` prefix seems better (and it's actually in the coding standards iirc)

File: projects/power_distribution/inc/current_measurement_config.h [hunk 3]
```diff
@@ -1,8 +1,8 @@
...
 
-// Standard hardware configurations for current_measurement.
+// Standard configurations for current_measurement.
 
 #include "current_measurement.h"
 
-extern const PowerDistributionCurrentHardwareConfig FRONT_POWER_DISTRIBUTION_CURRENT_HW_CONFIG;
-extern const PowerDistributionCurrentHardwareConfig REAR_POWER_DISTRIBUTION_CURRENT_HW_CONFIG;
+extern const CurrentMeasurementConfig FRONT_CURRENT_MEASUREMENT_CONFIG;
+extern const CurrentMeasurementConfig REAR_CURRENT_MEASUREMENT_CONFIG;
```
[the remark this answers is not in the document]
Comment (copperium) [reply projects/power_distribution/inc/current_measurement_config.h:8]: current_measurement is a bit long but I think isense is a bit too abbreviative, I'll change to current_sense

[the remark this answers is not in the document]
Comment (copperium) [reply projects/power_distribution/inc/current_measurement_config.h:8]: actually do you think that would be confusing with the current sense board in BMS? For now I'll leave it

File: libraries/ms-drivers/src/bts7200_load_switch.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (hewittmcg) [reply libraries/ms-drivers/src/bts7200_load_switch.c:185]: Yeah, I vaguely remember having a reason to have a max fault voltage when I changed up the BTS7200/7040 drivers, but I can't think of what it was now.  Per the datasheet, the fault current they output does have a max, but I don't see a reason why we wouldn't want to fault at a reading above that

File: projects/power_distribution/src/current_measurement.c [hunk 4]
```diff
@@ -1,129 +1,56 @@
...
+  // read from each output
+  for (uint8_t i = 0; i < s_config->num_outputs_to_read; i++) {
+    Output output = s_config->outputs_to_read[i];
+    if (output >= NUM_OUTPUTS) {
+      // shouldn't be possible, we caught them in init
+      continue;
+    }
+    StatusCode code = output_read_current(output, &s_storage.measurements[output]);
+    if (!status_ok(code)) {
+      LOG_WARN("Could not read current from output %d (index %d): code %d\n", output, i, code);
```
[the remark this answers is not in the document]
Comment (copperium) [reply projects/power_distribution/src/current_measurement.c:27]: GPIO outputs shouldn't be specified in the current measurement config - the abstraction is a bit leaky that way but I think it's ok to spam logs since we won't get valid data from those outputs and we don't want to tx bad data

File: projects/power_distribution/src/main.c [hunk 5]
```diff
@@ -32,17 +36,29 @@
 #define NUM_SIGNAL_BLINKS_BETWEEN_SYNCS 10
 #define VOLTAGE_REGULATOR_DELAY_MS 25
 
+// Like status_ok_or_return, but logs on failure for easier debugging.
+#define VERIFY_INIT(call)                                                                       \
```
[the remark this answers is not in the document]
Comment (copperium) [reply projects/power_distribution/src/main.c:40]: will do in a separate PR, this one's beefy enough as it is

File: projects/power_distribution/inc/current_measurement_config.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (jessm) [reply projects/power_distribution/inc/current_measurement_config.h]: yeah it's no biggie, makes sense to leave it, didn't think about conflicting with bms

**Review by jessm — APPROVED** [review]
Approval #2, merge at your leisure 👍
